### PR TITLE
(fix) colima: fix return code handling (followup to #3097)

### PR DESCRIPTION
**Pull request by tobitege**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

Fixes #3097 

- This fixes the exit code of the `colima start` command being passed to the outer loop.
- Added a 10 second delay before retry.
- Increased to 5 total attempts (4 retries).
- Added "--verbose" flag and lowered random port range to 10000-20000

### CODE REVIEW

**Comment by tobitege** [comment]
@SmartManoj This should work better now, right?

File: .github/workflows/run-unit-tests.yml [hunk 1]
```diff
@@ -89,17 +89,20 @@ jobs:
...
+            if ! colima start --network-address --arch x86_64 --cpu=1 --memory=1 --verbose --ssh-port $RANDOM_PORT; then
+              echo "Failed to start Colima."
+              return 1
+            fi
             docker context use $DOCKER_CONTEXT
           }
 
-          # Attempt to start Colima
-          ATTEMPT_LIMIT=3
+          # Attempt to start Colima; 4 retries:
```
Comment (SmartManoj) [inline .github/workflows/run-unit-tests.yml:104]:
```suggestion
          # Attempt to start Colima; 5 retries:
```

Comment (tobitege) [reply .github/workflows/run-unit-tests.yml:103]: Read the code: 5 total attempts, which means 4 retries.

File: .github/workflows/run-unit-tests.yml [hunk 2]
```diff
@@ -89,27 +89,32 @@ jobs:
...
+          ATTEMPT_LIMIT=5
           for ((i=1; i<=ATTEMPT_LIMIT; i++)); do
             if start_colima; then
               echo "Colima started successfully."
+              docker context use $DOCKER_CONTEXT
               break
             else
               echo "Failed to start Colima. Attempt $i/$ATTEMPT_LIMIT."
               if [ $i -eq $ATTEMPT_LIMIT ]; then
                 colima delete
```
Comment (SmartManoj) [inline .github/workflows/run-unit-tests.yml:113]: It should be executed before the last attempt.

[the remark this answers is not in the document]
Comment (tobitege) [reply .github/workflows/run-unit-tests.yml:113]: Why? The loop goes `i<=ATTEMPT_LIMIT`, so it should delete when last attempt failed.

[the remark this answers is not in the document]
Comment (SmartManoj) [reply .github/workflows/run-unit-tests.yml:113]: Deleting colima after the last attempt is no use.

[the remark this answers is not in the document]
Comment (tobitege) [reply .github/workflows/run-unit-tests.yml:113]: I think I can actually remove the delete completely, it is rather pointless now.

[the remark this answers is not in the document]
Comment (SmartManoj) [reply .github/workflows/run-unit-tests.yml:113]: It is a fallback method. https://github.com/OpenDevin/OpenDevin/pull/3061#issuecomment-2244160764

[the remark this answers is not in the document]
Comment (SmartManoj) [reply .github/workflows/run-unit-tests.yml:113]: > Meaning: I doubt that a delete between failed attempts has worked before.

delete between failed attempts itself has not happened before.

[the remark this answers is not in the document]
Comment (tobitege) [reply .github/workflows/run-unit-tests.yml]: I mean, we wouldn't _see_ that if all went well, but so far if colima failed, the delete didn't fix it 'cause then QEMU would've already terminated and what else happens then, unrecoverable state, I think.

[the remark this answers is not in the document]
Comment (SmartManoj) [reply .github/workflows/run-unit-tests.yml]: > delete between failed attempts itself has not happened before. [#3061 (files)](https://github.com/OpenDevin/OpenDevin/pull/3061/files#r1691249029)

colima is not started after delete

[the remark this answers is not in the document]
Comment (tobitege) [reply .github/workflows/run-unit-tests.yml]: So delete and re-install everytime? 
This isn't a fallback, that's just a hack.

[the remark this answers is not in the document]
Comment (SmartManoj) [reply .github/workflows/run-unit-tests.yml]: > So delete and re-install everytime?

No. deleting only for the last attempt

File: .github/workflows/run-unit-tests.yml [hunk 3]
```diff
@@ -87,35 +87,53 @@ jobs:
...
+              if [ -f /opt/homebrew/var/log/colima.log ]; then
+                cat /opt/homebrew/var/log/colima.log
               else
+                echo "Colima log file not found at /opt/homebrew/var/log/colima.log"
+                ls -l /opt/homebrew/var/log/
+              fi
+              if [ $i -lt $ATTEMPT_LIMIT ]; then
+                echo "Stopping Colima forcefully..."
                 colima stop -f
+                ecthe "ho "Waiting 10 seconds before next attempt..."
```
Comment (SmartManoj) [inline .github/workflows/run-unit-tests.yml:122]:
```suggestion
                echo "Waiting 10 seconds before next attempt..."
```

Comment (tobitege) [reply .github/workflows/run-unit-tests.yml]: 😂  yeh, just saw that

**Comment by tobitege** [comment]
@SmartManoj I've removed the "docker context" approach again.
Trying the loop with install and delete now. Also added that the IP will increase with every attempt.
From what I've seen thus far, it is the local IP address that it can't "resolve" or is blocked from using.
Bit of trial and error... 😔

**Comment by tobitege** [comment]
Oops a copy/paste error, one sec....

**Comment by tobitege** [comment]
> Github runner has only one IP. right?

I don't know, will see where it fails 😂

**Comment by SmartManoj** [comment]
> Github runner has only one IP. right?

For a single runner during execution. [Docs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#ip-addresses)

**Comment by tobitege** [comment]
> > Github runner has only one IP. right?
> 
> For a single runner during running. [Docs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#ip-addresses)

Thanks. I'll give it one more try with `limactl create colima-${IP_ADDRESS}` next and if that fails, will revert that IP changing as well.

**Comment by tobitege** [comment]
> > Github runner has only one IP. right?
> 
> For a single runner during running. [Docs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#ip-addresses)

But here we are using a local IP address only, not an external IP. So the docs don't apply in this case, I think.

File: .github/workflows/run-unit-tests.yml [hunk 4]
```diff
@@ -75,47 +75,56 @@ jobs:
...
-            echo "Using random port: $RANDOM_PORT for SSH"
+            # Find a free port in the range 10000-20000
+            RANDOM_PORT=$((RANDOM % 10001 + 10000))
 
-            colima start --network-address --arch x86_64 --cpu=1 --memory=1 --ssh-port $RANDOM_PORT
-            docker context use $DOCKER_CONTEXT
+            # Create Lima instance with the specific IP address
+            limactl create colima-${IP_ADDRESS}
+
+            if ! colima start --network-address $IP_ADDRESS --arch x86_64 --cpu=1 --memory=1 --verbose --ssh-port $RANDOM_PORT; then
```
Comment (SmartManoj) [inline .github/workflows/run-unit-tests.yml:91]: `network-address` is a boolean arg
https://github.com/abiosoft/colima/blob/c8c4c5a69b4e422dab76ac0a0f81515094302c2e/cmd/start.go#L160

Comment (tobitege) [reply .github/workflows/run-unit-tests.yml:91]: Again, you have to read the full source:
https://github.com/abiosoft/colima/blob/c8c4c5a69b4e422dab76ac0a0f81515094302c2e/cmd/start.go#L159-L161
```py
	if util.MacOS() {
		startCmd.Flags().BoolVar(&startCmdArgs.Network.Address, "network-address", false, "assign reachable IP address to the VM")
	}
```

**Comment by tobitege** [comment]
oh well, doesn't work with limactl neither, so going to remove that again ...

File: .github/workflows/run-unit-tests.yml (hunk 4 above)
Comment (SmartManoj) [reply .github/workflows/run-unit-tests.yml:91]: This line sets up a boolean command-line flag named network-address. When this flag is set to true by the user, it indicates that a reachable IP address should be assigned to the VM. By default, this flag is false. isn't it?

Comment (tobitege) [reply .github/workflows/run-unit-tests.yml:91]: 🤕 ohh, yes, I did just read the text, not the "BoolVar" and misunderstood it. 
There is not enough ☕☕☕around, apparently.
Thanks for correcting me there!

**Review by xingyaoww — APPROVED** [review]
Let's give it a try!

**Comment by WEI567** [comment]
可望打造新的

**Pull request merged**